### ISSUE TITLE
CVNN implementation (v18)

### ISSUE THREAD

**Issue by yabool2001** [yabool2001/temp#44]
bpsk_v0.1.17-rx_nn.py

CVNN (Complex-Valued Neural Networks)
1. Robimy natywne CVNN, ale na torch.complex64 (czyli torch.cfloat). Z perspektywy 12-bitowego PlutoSDR to i tak kosmiczny zapas precyzji, a zachowamy natywną matematykę zespoloną przy prędkości kilkudziesięciu Teraflopów na sekundę. NVIDIA z premedytacją blokuje sprzętowo wydajność FP64 na wszystkich kartach z serii konsumenckiej (GeForce).
2. X będzie zespolone, ale etykiety y to zwykłe klasy (np. 0, 1, 2), czyli torch.long.
3. w 3 linijkach napisać CReLU (Complex ReLU). Zwykłe F.relu() wyrzuci Ci błąd, bo na płaszczyźnie zespolonej z definicji nie istnieje pojęcie "mniejsze od zera" (co jest większe: $-1+2j$ czy $1-2j$?). 
4. Most przed LSTM: PyTorch ma gigantyczny problem z przepuszczaniem liczb zespolonych przez warstwy nn.LSTM. Dlaczego? Bo LSTM używa wewnątrz bramek opartych o tanh oraz sigmoid. Te funkcje trygonometryczne dla liczb urojonych posiadają tzw. osobliwości matematyczne i powodują, że błąd (loss) nagle staje się NaN (wybucha). Złotą zasadą AI-PHY jest użycie zespolonych splotów jako inteligentnych filtrów radiowych, a tuż przed pamięcią LSTM wyciągnięcie obwiedni sygnału (torch.abs).

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

# 1. WŁASNA AKTYWACJA ZESPOLONA
class ComplexReLU(nn.Module):
    def forward(self, z):
        # Aplikujemy standardowe ReLU oddzielnie na część rzeczywistą i urojoną,
        # odcinając ujemny szum z obu osi, po czym z powrotem sklejamy w complex.
        return torch.complex(F.relu(z.real), F.relu(z.imag))


# 2. ZESPOLONY ODBIORNIK NEURONOWY
class NativeComplexReceiver(nn.Module):
    def __init__(self, num_classes=3):
        super().__init__()
        
        # Wchodzi JEDEN kanał (cały zespolony ciąg I/Q jako jeden byt fizyczny)
        # dtype=torch.complex64 wymusza, by wagi wewnątrz sieci też były zespolone!
        self.conv1 = nn.Conv1d(
            in_channels=1, 
            out_channels=16, 
            kernel_size=8, 
            stride=4,       # Od razu robimy decymację dla Twojego SPS=4!
            padding=2,
            dtype=torch.complex64 
        )
        self.crelu = ComplexReLU()
        
        # Wracamy do świata rzeczywistego przed wejściem do pamięci
        self.lstm = nn.LSTM(
            input_size=16, 
            hidden_size=64, 
            batch_first=True, 
            dtype=torch.float32  # Zwykły float!
        )
        
        # Klasyfikator wypluwający szanse dla 3 klas (Szum, Marker, Payload)
        self.classifier = nn.Linear(64, num_classes)

    def forward(self, x):
        # x wchodzi jako: [Batch, Kanał=1, Czas=4096]
        
        x = self.conv1(x)
        x = self.crelu(x)
        
        # MAGIA CVNN: Po wykonaniu przestrzennego, zespolonego filtrowania fazy
        # wyciągamy obwiednię (EVM / moduł fizyczny). To płynnie konwertuje 
        # wynik z complex64 z powrotem na float32 i ratuje LSTM przed NaNs!
        x = torch.abs(x) 
        
        # Zmiana wymiarów pod wlot do LSTM: z [Batch, Cechy, Czas] na [Batch, Czas, Cechy]
        x = x.transpose(1, 2)
        
        x, (h_n, c_n) = self.lstm(x)
        
        # Wypluwamy surowe szanse dla klas na całej osi czasu
        logits = self.classifier(x)
        
        # Zwraca: [Batch, 1024, 3] (1024, bo zrobiliśmy stride=4)
        return logits


# ==========================================
# 3. TEST ZDERZENIOWY DLA TWOJEGO RTX 5080
# ==========================================
if __name__ == "__main__":
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print(f"Silnik AI ustawiony na: {device}")
    
    # Symulacja zrzutu z Twojej funkcji: 128 buforów, 1 zespolony kanał, długość 4096
    # TO JEST TWOJE X_train
    dummy_X = torch.randn(128, 1, 4096, dtype=torch.complex64).to(device)
    
    # Symulacja etykiet (Nauczyciela): Klasy od 0 do 2
    # Uwaga: czas ma długość 1024, bo warstwa conv1(stride=4) go 4-krotnie skróciła
    dummy_y = torch.randint(0, 3, (128, 1024), dtype=torch.long).to(device)
    
    model = NativeComplexReceiver().to(device)
    
    # Przepuszczamy sygnał (Forward pass)
    wyjscie = model(dummy_X)
    
    print(f"Zespolone wejście X : {dummy_X.shape} | Typ: {dummy_X.dtype}")
    print(f"Rzeczywiste Wyjście : {wyjscie.shape} | Typ: {wyjscie.dtype}")
    
    # Liczymy stratę! (PyTorch oczekuje wymiaru klas na drugiej pozycji, dlatego transpose)
    loss_fn = nn.CrossEntropyLoss()
    strata = loss_fn(wyjscie.transpose(1, 2), dummy_y)
    
    print(f"Strata na starcie = {strata.item():.4f}")
    print("ZADZIAŁAŁO! Czerstwiak odszczekuje spod stołu! 🐶")
```

**Comment by yabool2001** [comment]
udało się 40edee6dbe35b00b368eea00f66ae717ac0f0cf2
